File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root == null){
            return null;
        }
        
        Queue<Node> q = new ArrayDeque<>();
        
        Node prev_n = null;
        q.add(root);
        
        while(!q.isEmpty()){
            int cnt = q.size();
            for (int i = 0; i < cnt; i++){
                Node node = q.poll();
                if (prev_n != null){
                    prev_n.next = node;
                }
                
                prev_n = node;
                
                if (node.left != null){
                    q.add(node.left);
                }
                if (node.right != null){
                    q.add(node.right);
                }
                
                if (i + 1 == cnt){
                    prev_n = null;
                }
            }
        }
        
        return root;
    }
}

File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (!root) {
            return nullptr;
        }
        
        connectNodeRec(root->left, root->right);

	    return root;
    }
    
    void connectNodeRec(Node* node1, Node* node2) {
        if (!node1) {
            return;
        }

        node1->next = node2;

        connectNodeRec(node1->left, node1->right);
        connectNodeRec(node2->left, node2->right);
        connectNodeRec(node1->right, node2->left);
    }
};